drivers/watchdog/wdt_rpi_pico.c

```diff
@@ -0,0 +1,174 @@
+			     WATCHDOG_CTRL_PAUSE_DBG1_BITS));
+	}
+
+	watchdog_hw->load = data->load;
+
+	/* Zero out the scratch registers so that the module reboots at the
+	 * default program counter
+	 */
+	watchdog_hw->scratch[4] = 0;
+	watchdog_hw->scratch[5] = 0;
+	watchdog_hw->scratch[6] = 0;
+	watchdog_hw->scratch[7] = 0;
+
+	hw_set_bits(&watchdog_hw->ctrl, WATCHDOG_CTRL_ENABLE_BITS);
+
+	data->enabled = true;
+
+	watchdog_hw->tick = (config->xtal_frequency / RPI_2040_XTAL_FREQ_WDT_TICK_DIVISOR) |
+			    WATCHDOG_TICK_ENABLE_BITS;
+
+	return 0;
+}
+
+static int wdt_rpi_pico_disable(const struct device *dev)
+{
+	struct wdt_rpi_pico_data *data = dev->data;
+
+	if (data->enabled == false) {
+		return -EFAULT;
+	}
+
+	hw_clear_bits(&watchdog_hw->ctrl, WATCHDOG_CTRL_ENABLE_BITS);
+
+	data->enabled = false;
+
+	return 0;
+}
+
+static int wdt_rpi_pico_install_timeout(const struct device *dev, const struct wdt_timeout_cfg *cfg)
+{
+	struct wdt_rpi_pico_data *data = dev->data;
+
+	if (cfg->window.min != 0U || cfg->window.max == 0U) {
+		return -EINVAL;
+	} else if (cfg->window.max > RPI_2040_MAX_WDT_TIME) {
+		return -EINVAL;
+	} else if (cfg->callback != NULL) {
+		return -EINVAL;
+	} else if ((cfg->flags & WDT_FLAG_RESET_MASK) == WDT_FLAG_RESET_NONE) {
+		/* The RP2040 does technically support this mode, but requires
+		 * a program counter and stack pointer value to be set,
+		 * therefore do not allow configuring in this mode
+		 */
+		return -EINVAL;
+	}
+
+	data->load = (cfg->window.max * RPI_2020_WDT_TIME_MULTIPLCATION_FACTOR);
+	data->reset_type = (cfg->flags & WDT_FLAG_RESET_MASK);
+
+	return 0;
+}
+
+static int wdt_rpi_pico_feed(const struct device *dev, int channel_id)
+{
+	struct wdt_rpi_pico_data *data = dev->data;
+
+	if (channel_id != 0) {
+		/* There is only one input to the watchdog */
+		return -EINVAL;
+	}
+
+	if (data->enabled == false) {
+		/* Watchdog is not running so does not need to be fed */
+		return -EINVAL;
+	}
+
+	watchdog_hw->load = data->load;
+
+	return 0;
+}
+
+static int wdt_rpi_pico_init(const struct device *dev)
+{
+#ifndef CONFIG_WDT_DISABLE_AT_BOOT
+	return wdt_rpi_pico_setup(dev, WDT_OPT_PAUSE_HALTED_BY_DBG);
+#endif
+
+	return 0;
+}
+
+static const struct wdt_driver_api wdt_rpi_pico_driver_api = {
+	.setup = wdt_rpi_pico_setup,
+	.disable = wdt_rpi_pico_disable,
+	.install_timeout = wdt_rpi_pico_install_timeout,
+	.feed = wdt_rpi_pico_feed,
+};
+
+#define WDT_RPI_PICO_WDT_DEVICE(idx)                                                               \
+	static const struct wdt_rpi_pico_config wdt_##idx##_config = {                             \
+		.xtal_frequency = DT_INST_PROP_BY_PHANDLE(idx, clocks, clock_frequency)            \
+	};                                                                                         \
+	static struct wdt_rpi_pico_data wdt_##idx##_data = {                                       \
+		.reset_type = WDT_FLAG_RESET_SOC,                                                  \
```

Comment:
Shouldn't `reset_type` and `load` be part of the config, not the data?

Reply:
These are runtime settable when calling the `install_timeout` function so cannot be part of the const config